code/game/gamemodes/miniantags/pulsedemon/pulsedemon.dm

```diff
@@ -0,0 +1,713 @@
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/toggle/do_drain(do_drain))
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/toggle/can_exit_cable(can_exit_cable))
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/cablehop)
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/emagtamper)
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/emp)
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/overload)
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/remotehijack)
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/remotedrain)
+	AddSpell(new /obj/effect/proc_holder/spell/pulse_demon/open_upgrades)
+
+/mob/living/simple_animal/pulse_demon/Stat()
+	. = ..()
+	if(statpanel("Status"))
+		stat(null, "Charge: [format_si_suffix(charge)]W")
+		stat(null, "Maximum Charge: [format_si_suffix(maxcharge)]W")
+		stat(null, "Drained Charge: [format_si_suffix(charge_drained)]W")
+		stat(null, "Hijacked APCs: [length(hijacked_apcs)]")
+		stat(null, "Drain Rate: [format_si_suffix(power_drain_rate)]W")
+		stat(null, "Hijack Time: [hijack_time / 10] seconds")
+
+/mob/living/simple_animal/pulse_demon/dust()
+	return death()
+
+/mob/living/simple_animal/pulse_demon/gib()
+	return death()
+
+/mob/living/simple_animal/pulse_demon/death()
+	pb_helper.cancel() // clean up any actions we were doing
+	var/turf/T = get_turf(src)
+	do_sparks(rand(2, 4), FALSE, src)
+	. = ..()
+
+	var/heavy_radius = min(charge / 50000, 20)
+	var/light_radius = min(charge / 25000, 25)
+	empulse(T, heavy_radius, light_radius)
+	playsound(T, pick(hurt_sounds), 50, TRUE)
+
+/mob/living/simple_animal/pulse_demon/proc/exit_to_turf(atom/oldloc)
+	var/turf/T = get_turf(loc)
+	current_power = null
+	update_controlling_area()
+	current_cable = null
+	forceMove(T)
+	Move(T)
+	if(!current_cable && !current_power)
+		var/obj/effect/proc_holder/spell/pulse_demon/toggle/can_exit_cable/S = locate() in mob_spell_list
+		if(!S.locked && !can_exit_cable)
+			can_exit_cable = TRUE
+			S.do_toggle(can_exit_cable)
+
+/mob/living/simple_animal/pulse_demon/proc/update_controlling_area(reset = FALSE)
+	var/area/prev = controlling_area
+	if(reset || current_power == null)
+		controlling_area = null
+	else
+		var/obj/machinery/power/apc/A = current_power
+		if(istype(A) && (A in hijacked_apcs))
+			controlling_area = A.apc_area
+		else
+			controlling_area = null
+
+	if((prev == null && controlling_area == null) || (prev != null && controlling_area != null))
+		return // only update icons when we get or no longer have ANY area
+	for(var/obj/effect/proc_holder/spell/pulse_demon/S in mob_spell_list)
+		if(!S.action || S.locked)
+			continue
+		if(S.requires_area)
+			S.action.UpdateButtonIcon()
+
+// can enter an apc at all?
+/mob/living/simple_animal/pulse_demon/proc/is_valid_apc(obj/machinery/power/apc/A)
+	return istype(A) && !(A.stat & BROKEN) && !A.shorted
+
+/mob/living/simple_animal/pulse_demon/Move(newloc)
```

Comment:
Trying to revert movement if a pulse_demon_enter or some equivalent returned false seems to me like it would introduce rubber banding (not to mention it'd compound the jumpy forceMove problem pointed out by Henri), and if we keep the locate(...)s in Move we've already done the expensive part since the rest is just checks and procedures we'd do regardless in any pulse_demon_enter proc.
I don't see a way to do a smoother implementation here, but I'm not an expert on byond movement code.
I could make it only try to locate a cable if new_power remains non-null after its two checks, which would be a bit better, at least.